Add PDC network and PDC network token resources

aangelisc opened the pull request:
Feedback welcome here (my first time creating some TF resources), this is essentially a duplicate of the access policy and token resources with some minor changes that are specific to PDC.

Fixes #1222 
Fixes #1638
Fixes https://github.com/grafana/hosted-grafana/issues/4278

aangelisc committed:
Add PDC network and PDC network token resources

- Add resource lister function that will return a slice of resources

aangelisc committed:
Support PDC in data sources

aangelisc committed:
Add examples and generate docs

aangelisc committed:
Regen

dafydd-t commented on examples/resources/grafana_cloud_private_datasource_connect/resource.tf:
```diff
@@ -0,0 +1,17 @@
+  stackID = "<your stack ID>"
+}
+
+resource "grafana_cloud_private_datasource_connect" "test" {
```
This is the PDC network, I'd like to be explicit with naming
```suggestion
resource "grafana_cloud_private_datasource_connect_network" "test" {
```

aangelisc replied:
Done!

dafydd-t commented on examples/resources/grafana_cloud_private_datasource_connect/resource.tf:
```diff
@@ -0,0 +1,17 @@
+  region       = "us"
+  name         = "my-pdc"
+  display_name = "My PDC"
+  identifier   = data.grafana_cloud_stack.current.stackID
```
I'm not sure of conventions here, but I'd prefer if this name was more explicit:
```suggestion
  stack_identifier   = data.grafana_cloud_stack.current.stackID
```

aangelisc replied:
Done!

dafydd-t commented on examples/resources/grafana_cloud_private_datasource_connect/resource.tf:
```diff
@@ -0,0 +1,17 @@
+
+resource "grafana_cloud_private_datasource_connect_token" "test" {
+  pdc_network_id = grafana_cloud_private_datasource_connect.test.network_id
+  region         = "us"
```
could we pass the whole network resource as input? Then we could get the region from the network schema, and we wouldn't need to repeat the `region` input in the token schema. This would feel more natural to me, since a token MUST be in the same region as its Access Policy

aangelisc replied:
I don't think this is straightforward. I've updated the example to use the region from the previously created network.

dafydd-t replied:
Fair enough. I'll leave this comment open in case someone with some TF experience has input.

dafydd-t replied:
I also see that this follows the same practice as the existing access policy token terraform, the consistency is more important 👍

dafydd-t commented on internal/resources/cloud/resource_private_datasource_connect.go:
```diff
@@ -0,0 +1,253 @@
+				Type:         schema.TypeString,
+				Required:     true,
+				ForceNew:     true,
+				Description:  "Region where the API is deployed. Generally where the stack is deployed. Use the region list API to get the list of available regions: https://grafana.com/docs/grafana-cloud/developer-resources/api-reference/cloud-api/#list-regions.",
```
I feel like this should be more certain for our users:

```suggestion
				Description:  "The region where your stack is deployed. Use the instances list API to get the region for your instance - use the regionSlug property: https://grafana.com/docs/grafana-cloud/developer-resources/api-reference/cloud-api/#list-stacks",
```

aangelisc replied:
Done!

dafydd-t commented on internal/resources/cloud/resource_private_datasource_connect.go:
```diff
@@ -0,0 +1,253 @@
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Name of the PDC network.",
```
Can we describe the casing requirements here? We need the string to be hyphenated, lowercase

aangelisc replied:
Are these requirements documented anywhere for us to refer to?

dafydd-t replied:
Took a peek - the access policies requirements are [here](https://grafana.com/docs/grafana-cloud/developer-resources/api-reference/cloud-api/#request-body)

dafydd-t commented on internal/resources/grafana/resource_data_source.go:
```diff
@@ -119,6 +119,12 @@ source selected (via the 'type' argument).
 			},
 			"json_data_encoded":        datasourceJSONDataAttribute(),
 			"secure_json_data_encoded": datasourceSecureJSONDataAttribute(),
+			"private_datasource_connect_id": {
```
```suggestion
			"private_datasource_connect_network_id": {
```

aangelisc replied:
Done!

dafydd-t reviewed (changes requested):
This looks really good! Thanks a lot for starting it off. I've added some comments around the naming of things for now. We need a more thorough review of the terraform itself, we can get to that soon.

aangelisc committed:
Update data source property name

aangelisc committed:
Add PDC network data sources

aangelisc committed:
Rename resources for clarity

- Tidy up property names

aangelisc commented:
Updated based on your suggestions @dafydd-t. I've also added a data source to allow retrieval of networks that may not be managed by TF.

aangelisc committed:
Some more naming updates

aangelisc committed:
Update docs

aangelisc committed:
Merge branch 'main' into andreas/pdc-provider

aangelisc committed:
Fix pdc network Id check

aangelisc committed:
Update docs and add test

aangelisc committed:
Docs update

aangelisc committed:
Add reference to access policies naming reqs

dafydd-t commented on docs/data-sources/data_source.md:
```diff
@@ -59,6 +59,7 @@ data "grafana_data_source" "from_uid" {
 - `id` (String) The ID of this resource.
 - `is_default` (Boolean) Whether to set the data source as default. This should only be `true` to a single data source.
 - `json_data_encoded` (String) Serialized JSON string containing the json data. This attribute can be used to pass configuration options to the data source. To figure out what options a datasource has available, see its docs or inspect the network data when saving it from the Grafana UI. Note that keys in this map are usually camelCased.
+- `private_datasource_connect_network_id` (String) (Can only be used with data sources in Grafana Cloud) The ID of the private data source network to use with this data source.
```
```suggestion
- `private_datasource_connect_network_id` (String) (Can only be used with data sources in Grafana Cloud) The ID of the Private Data source Connect network to use with this data source.
```

aangelisc replied:
Fixed

dafydd-t commented on docs/data-sources/cloud_private_data_source_connect_networks.md:
```diff
@@ -0,0 +1,47 @@
+page_title: "grafana_cloud_private_data_source_connect_networks Data Source - terraform-provider-grafana"
+subcategory: "Cloud"
+description: |-
+  Fetches Private data source connect networks from Grafana Cloud.
```
```suggestion
  Fetches Private Data source Connect networks from Grafana Cloud.
```

dafydd-t replied:
doh i see this is generated, but please update in the source

aangelisc replied:
Fixed in source

dafydd-t commented on docs/data-sources/cloud_private_data_source_connect_networks.md:
```diff
@@ -0,0 +1,47 @@
+
+# grafana_cloud_private_data_source_connect_networks (Data Source)
+
+Fetches Private data source connect networks from Grafana Cloud.
```
```suggestion
Fetches Private Data source Connect networks from Grafana Cloud.
```

aangelisc replied:
Fixed

dafydd-t commented on internal/resources/cloud/data_source_private_data_source_connect_network.go:
```diff
@@ -0,0 +1,130 @@
+	)
+}
+
+type PrivateDatasourceConnectNetworksDataSource struct {
```
Can we be consistent about typing on `Datasource` please?

dafydd-t replied:
I think it would be OK to use PDC in the type names too - up to you.

aangelisc replied:
Switched to using PDC

dafydd-t commented:
@aangelisc  do you have an example terraform plan that you used to run this locally? I'd love to give it a go

aangelisc committed:
Minor updates for type names and docs

aangelisc commented:
Hey @dafydd-t, I've pushed some more updates 😊

You should be able to test this with the following tf:

```tf
terraform {
  required_providers {
    grafana = {
      source = "grafana/grafana"
    }
  }
}

provider "grafana" {
  alias                     = "cloud"
  cloud_access_policy_token = "<token>"
}


data "grafana_cloud_stack" "stack" {
  provider = grafana.cloud

  slug = "stack"
}

resource "grafana_cloud_stack_service_account" "cloud_sa" {
  provider   = grafana.cloud
  stack_slug = data.grafana_cloud_stack.stack.slug

  name        = "pdc-test-sa"
  role        = "Admin"
  is_disabled = false
}

resource "grafana_cloud_stack_service_account_token" "cloud_sa" {
  provider   = grafana.cloud
  stack_slug = data.grafana_cloud_stack.stack.slug

  name               = "terraform serviceaccount key"
  service_account_id = grafana_cloud_stack_service_account.cloud_sa.id
}

provider "grafana" {
  alias = "stack"

  url                       = data.grafana_cloud_stack.stack.url
  auth                      = grafana_cloud_stack_service_account_token.cloud_sa.key
  cloud_access_policy_token = "<token>"
}

resource "grafana_cloud_private_data_source_connect_network" "test" {
  provider = grafana.stack

  region           = data.grafana_cloud_stack.stack.region_slug
  name             = "test-pdc"
  display_name     = "Test PDC"
  stack_identifier = data.grafana_cloud_stack.stack.id
}

resource "grafana_cloud_private_data_source_connect_network_token" "test" {
  provider = grafana.stack

  pdc_network_id = grafana_cloud_private_data_source_connect_network.test.pdc_network_id
  region         = grafana_cloud_private_data_source_connect_network.test.region
  name           = "my-pdc-token"
  display_name   = "My PDC Token"
}

resource "grafana_data_source" "influxdb" {
  provider = grafana.stack

  type                = "influxdb"
  name                = "pdc-test-influx"
  url                 = "http://influxdb.example.net:8086/"
  basic_auth_enabled  = true
  basic_auth_username = "username"
  database_name       = "dbname" // Example: influxdb_database.metrics.name

  json_data_encoded = jsonencode({
    authType          = "default"
    basicAuthPassword = "mypassword"
  })

  private_data_source_connect_network_id = grafana_cloud_private_data_source_connect_network.test.pdc_network_id
}

```

aangelisc committed:
Fix replace

aangelisc committed:
Merge branch 'main' into andreas/pdc-provider

dafydd-t commented:
Thanks - I tried this  myself to create a cloud stack, PDC network and some tokens and it works a treat!

One behaviour i don't understand is when I run `terraform apply` with no changes, I see this:


```
# grafana_data_source.influxdb will be updated in-place
  ~ resource "grafana_data_source" "influxdb" {
        id                                     = "1:be9kz8qdpwbnka"
      ~ json_data_encoded                      = jsonencode(
          ~ {
              - enableSecureSocksProxy   = true
              - secureSocksProxyUsername = "99d012ce-4db8-4810-887f-90479122d333"
                # (2 unchanged attributes hidden)
            }
        )
        name                                   = "pdc-test-influx"
        # (11 unchanged attributes hidden)
    }
```

Why are these properties in the datasource getting deleted? But also, when i apply it and go check the datasource, it still looks configured correctly :/

Apart from that, from a PDC perspective this looks good! I've asked again in #platform for a terraform focused review.

spinillos commented on internal/common/resource.go:
```diff
@@ -72,6 +75,7 @@ type Resource struct {
 
 	// Generation configuration
 	ListIDsFunc                ResourceListIDsFunc
+	ListFunc                   ResourceListFunc
```
I saw that you created a new value and a function to set it, but it isn't used, right?

aangelisc replied:
This is actually used [here](https://github.com/grafana/terraform-provider-grafana/blob/ca81e8d9440705a7e8c87352290815ae25ac7d15/internal/resources/cloud/resource_private_data_source_connect_network.go#L102).

It's not necessarily a requirement but I think it's a nice feature to be able to list all the resources of a specific type. If we don't want this then we can remove it 😊

spinillos replied:
For example, `ListIDsFunc` is used [here](https://github.com/grafana/terraform-provider-grafana/blob/main/pkg/generate/generate.go#L207) and other parts for testing, but this new `LisFunc` is set but no one is calling it.

aangelisc replied:
That's fair, I'll remove this 😊

aangelisc committed:
Fix warning

aangelisc committed:
Suppress pdc diffs

aangelisc commented:
> Thanks - I tried this myself to create a cloud stack, PDC network and some tokens and it works a treat!
> 
> One behaviour i don't understand is when I run `terraform apply` with no changes, I see this:
> 
> ```
> # grafana_data_source.influxdb will be updated in-place
>   ~ resource "grafana_data_source" "influxdb" {
>         id                                     = "1:be9kz8qdpwbnka"
>       ~ json_data_encoded                      = jsonencode(
>           ~ {
>               - enableSecureSocksProxy   = true
>               - secureSocksProxyUsername = "99d012ce-4db8-4810-887f-90479122d333"
>                 # (2 unchanged attributes hidden)
>             }
>         )
>         name                                   = "pdc-test-influx"
>         # (11 unchanged attributes hidden)
>     }
> ```
> 
> Why are these properties in the datasource getting deleted? But also, when i apply it and go check the datasource, it still looks configured correctly :/
> 
> Apart from that, from a PDC perspective this looks good! I've asked again in #platform for a terraform focused review.

I've fixed this! It's because the data source `jsonData` value for the diff is determined before the actual final value is calculated. I've updated this to suppress the diff if the `private_data_source_connect_network_id` is set by also setting the relevant `jsonData` properties.

aangelisc committed:
Remove resource lister

aangelisc committed:
Remove type

dafydd-t commented:
That issue looks fixed! Thanks. I'm still seeing one unexpected behaviour. I've got three tokens in my sample terraform, and running `terraform apply` with no changes, I see this diff for  one of them. This one has an expiry date, which doesnt change! but the diff thinks it does:

```
Terraform will perform the following actions:

  # grafana_cloud_private_data_source_connect_network_token.test3 must be replaced
-/+ resource "grafana_cloud_private_data_source_connect_network_token" "test3" {
      ~ created_at     = "2025-01-16T10:21:45Z" -> (known after apply)
      ~ expires_at     = "2025-02-01T00:00:00Z" -> "2025-02-01T00:00:00.00Z" # forces replacement
      ~ id             = "prod-gb-south-0:0a466813-3e6d-4e0d-ac09-2c6ba70b7669" -> (known after apply)
        name           = "my-pdc-token-3"
      ~ token          = (sensitive value)
      + updated_at     = (known after apply)
        # (3 unchanged attributes hidden)
    }

Plan: 1 to add, 0 to change, 1 to destroy.
```


I think once we've fixed this one we're good to go.

aangelisc committed:
Merge branch 'main' into andreas/pdc-provider

spinillos reviewed (approved):
The code looks like good to me

dafydd-t reviewed (approved):
Thank you!
